File: crm-backend/src/main/java/com/arjun/crm/service/EmailCampaignService.java
package com.arjun.crm.service;

import com.arjun.crm.dto.request.*;
import com.arjun.crm.dto.response.EmailCampaignResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * EmailCampaignService - FEATURE #3
 * 
 * Service interface for email campaign management
 * 
 * Permission Model:
 * - OWNER/ADMIN: Full access (create, read, update, delete, schedule, send)
 * - MEMBER: Read-only access (list, get, view analytics)
 * 
 * All operations are workspace-scoped through WorkspaceAuthorizationService
 */
public interface EmailCampaignService {
    
    /**
     * Create a new email campaign
     * 
     * Permission: OWNER/ADMIN only
     * 
     * @param workspaceId the workspace ID
     * @param request creation request
     * @return created campaign response
     * @throws com.arjun.crm.exception.AccessDeniedException if insufficient permission
     * @throws com.arjun.crm.exception.ResourceNotFoundException if workspace not found
     */
    EmailCampaignResponse createCampaign(Long workspaceId, CreateEmailCampaignRequest request);
    
    /**
     * List all campaigns in a workspace (paginated)
     * 
     * Permission: Any workspace member
     * 
     * @param workspaceId the workspace ID
     * @param pageable pagination settings
     * @return page of campaigns
     * @throws com.arjun.crm.exception.AccessDeniedException if not workspace member
     */
    Page<EmailCampaignResponse> listCampaigns(Long workspaceId, Pageable pageable);
    
    /**
     * Get a specific campaign by ID
     * 
     * Permission: Any workspace member
     * 
     * @param workspaceId the workspace ID
     * @param campaignId the campaign ID
     * @return campaign response
     * @throws com.arjun.crm.exception.AccessDeniedException if not workspace member
     * @throws com.arjun.crm.exception.ResourceNotFoundException if campaign not found
     */
    EmailCampaignResponse getCampaign(Long workspaceId, Long campaignId);
    
    /**
     * List campaigns by status
     * 
     * Permission: Any workspace member
     * 
     * @param workspaceId the workspace ID
     * @param status campaign status
     * @param pageable pagination settings
     * @return page of campaigns
     */
    Page<EmailCampaignResponse> listCampaignsByStatus(Long workspaceId, String status, Pageable pageable);
    
    /**
     * Update campaign (DRAFT only)
     * 
     * Permission: OWNER/ADMIN only
     * 
     * @param workspaceId the workspace ID
     * @param campaignId the campaign ID
     * @param request update request
     * @return updated campaign response
     * @throws com.arjun.crm.exception.AccessDeniedException if insufficient permission
     * @throws com.arjun.crm.exception.ResourceNotFoundException if campaign not found
     * @throws IllegalStateException if campaign status is not DRAFT
     */
    EmailCampaignResponse updateCampaign(Long workspaceId, Long campaignId, UpdateEmailCampaignRequest request);
    
    /**
     * Schedule campaign for sending
     * 
     * Permission: OWNER/ADMIN only
     * 
     * @param workspaceId the workspace ID
     * @param campaignId the campaign ID
     * @param request schedule request
     * @return updated campaign response
     * @throws com.arjun.crm.exception.AccessDeniedException if insufficient permission
     * @throws com.arjun.crm.exception.ResourceNotFoundException if campaign not found
     */
    EmailCampaignResponse scheduleCampaign(Long workspaceId, Long campaignId, ScheduleCampaignRequest request);
    
    /**
     * Send campaign immediately
     * 
     * Permission: OWNER/ADMIN only
     * 
     * @param workspaceId the workspace ID
     * @param campaignId the campaign ID
     * @return updated campaign response with status SENDING
     * @throws com.arjun.crm.exception.AccessDeniedException if insufficient permission
     */
    EmailCampaignResponse sendCampaign(Long workspaceId, Long campaignId);
    
    /**
     * Update campaign status
     * 
     * Permission: OWNER/ADMIN only
     * 
     * @param workspaceId the workspace ID
     * @param campaignId the campaign ID
     * @param request status request
     * @return updated campaign response
     */
    EmailCampaignResponse updateCampaignStatus(Long workspaceId, Long campaignId, EmailCampaignStatusRequest request);
    
    /**
     * Delete campaign (soft delete with deleted_at)
     * 
     * Permission: OWNER/ADMIN only
     * 
     * @param workspaceId the workspace ID
     * @param campaignId the campaign ID
     * @throws com.arjun.crm.exception.AccessDeniedException if insufficient permission
     */
    void deleteCampaign(Long workspaceId, Long campaignId);
}

File: crm-frontend/src/services/emailCampaignService.js
import api from './api'

const unwrap = (response) => response?.data ?? response
const campaignPath = (workspaceId, campaignId = '') =>
  `/workspaces/${workspaceId}/email-campaigns${campaignId ? `/${campaignId}` : ''}`

export const emailCampaignService = {
  async listCampaigns(workspaceId, { page = 0, size = 20, sortBy = 'createdAt', status } = {}) {
    const path = status ? `${campaignPath(workspaceId)}/status/${status}` : campaignPath(workspaceId)
    return unwrap(await api.get(path, { params: { page, size, sortBy } }))
  },
  async getCampaign(workspaceId, campaignId) {
    return unwrap(await api.get(campaignPath(workspaceId, campaignId)))
  },
  async createCampaign(workspaceId, payload) {
    return unwrap(await api.post(campaignPath(workspaceId), payload))
  },
  async updateCampaign(workspaceId, campaignId, payload) {
    return unwrap(await api.put(campaignPath(workspaceId, campaignId), payload))
  },
  async deleteCampaign(workspaceId, campaignId) {
    return unwrap(await api.delete(campaignPath(workspaceId, campaignId)))
  },
  async sendCampaign(workspaceId, campaignId) {
    return unwrap(await api.post(`${campaignPath(workspaceId, campaignId)}/send`))
  },
  async scheduleCampaign(workspaceId, campaignId, payload) {
    return unwrap(await api.post(`${campaignPath(workspaceId, campaignId)}/schedule`, payload))
  },
  async listRecipients(workspaceId, campaignId, { page = 0, size = 50, sortBy = 'createdAt' } = {}) {
    return unwrap(await api.get(`${campaignPath(workspaceId, campaignId)}/recipients`, { params: { page, size, sortBy } }))
  },
  async addRecipients(workspaceId, campaignId, payload) {
    return unwrap(await api.post(`${campaignPath(workspaceId, campaignId)}/recipients`, payload))
  },
  async removeRecipient(workspaceId, campaignId, recipientId) {
    return unwrap(await api.delete(`${campaignPath(workspaceId, campaignId)}/recipients/${recipientId}`))
  },
  async listTemplates(workspaceId, { page = 0, size = 100, sortBy = 'createdAt' } = {}) {
    return unwrap(await api.get(`/workspaces/${workspaceId}/email-templates`, { params: { page, size, sortBy } }))
  },
  async listSegments(workspaceId, { page = 0, size = 100, sortBy = 'createdAt' } = {}) {
    return unwrap(await api.get(`/workspaces/${workspaceId}/email-segments`, { params: { page, size, sortBy } }))
  },
  async createTemplate(workspaceId, payload) {
    return unwrap(await api.post(`/workspaces/${workspaceId}/email-templates`, payload))
  },
}
